File: max_subarray/max_sub.js
function maxSub(nums){
  let maxSum = nums[0]
  let sum    = nums[0]

  for (i = 1; i < nums.length; i++){
    sum = Math.max(nums[i], nums[i]+sum)
    maxSum = Math.max(maxSum, sum)
  }
  return maxSum
}


console.log(maxSub([5,4,-1,7,8]))

File: max_subarray/max_sub.rb
def max_sub(nums)
  current_sum = nums[0]
  max_sum     = nums[0]

  nums[1..nums.length].each do |num|
    current_sum = [current_sum + num, num].max
    max_sum     = [max_sum, current_sum].max
  end
  max_sum
end

p max_sub([5,4,-1,7,8])
